iree/samples/variables_and_state/main.c

```diff
@@ -0,0 +1,182 @@
+                                           &instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_create(
+      &instance_options, iree_allocator_system(), &instance));
+  // TODO(#5724): move device selection into the compiled modules.
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_try_create_default_device(
+      instance, driver_name, &device));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Session configuration (one per loaded module to hold module state).
+  fprintf(stdout, "Creating IREE runtime session\n");
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_session_create_with_device(
+      instance, &session_options, device,
+      iree_runtime_instance_host_allocator(instance), &session));
+  iree_hal_device_release(device);
+
+  fprintf(stdout, "Loading bytecode module at '%s'\n",
+          bytecode_module_path.data);
+  IREE_RETURN_IF_ERROR(iree_runtime_session_append_bytecode_module_from_file(
+      session, bytecode_module_path.data));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Call functions.
+  fprintf(stdout, "Calling functions\n\n");
+  iree_runtime_call_t call_get_value;
+  iree_runtime_call_t call_set_value;
+  iree_runtime_call_t call_add_to_value;
+  iree_runtime_call_t call_reset_value;
+  // Exported function names match what we used in Python.
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.get_value"), &call_get_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.set_value"), &call_set_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.add_to_value"),
+      &call_add_to_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.reset_value"),
+      &call_reset_value));
+
+  iree_hal_buffer_view_t* ret0 = NULL;
+
+  // Call get_value() to check the initial value.
+  IREE_RETURN_IF_ERROR(iree_runtime_call_invoke(&call_get_value, /*flags=*/0));
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_outputs_pop_front_buffer_view(&call_get_value, &ret0));
+  fprintf(stdout, "Initial value from get_value():\n");
+  IREE_RETURN_IF_ERROR(
+      iree_hal_buffer_view_fprint(stdout, ret0, /*max_element_count=*/64));
+
+  // Call set_value() then get_value() again.
+  iree_hal_buffer_view_t* arg0 = NULL;
```

Comment:
another approach here is to mimic the structure of the original module - this would make the code much easier to read and avoid the variable reuse that makes it harder to reuse chunks:

```c
iree_hal_buffer_view_t* initial_value = NULL;
IREE_RETURN_IF_ERROR(counter_get_value(&call_get_value, &initial_value));
IREE_RETURN_IF_ERROR(iree_hal_buffer_view_fprint(stdout, initial_value, /*max_element_count=*/64));
iree_hal_buffer_view_release(initial_value);
```

And it'd be even cooler if we auto generated those functions from the module - could even be a little tool that ran on the vmfb. But I don't think having them as hand-written functions above would be out of place here and I don't see much positive value of having all these calls just floating in the same scope.

Reply:
it's also much harder to observe lifetime here - you're missing a release of ret0 above for example :)

Reply:
Generating from the vmfb would be nice. I ran into some lifetime issues with stack local variables that I can't reasonably debug at the moment. I worked around those by cloning instead of wrapping. Maybe you could help take a look?

Reply:
I don't spot anything immediately, but may spend some time fiddling with it this weekend.